File: src/lib/pipeline/parsers/nb-parser.ts
/**
 * nb-parser.ts — Client-side port of nb_parser.py
 *
 * Handles two Excel formats:
 *   1. Workbook export — clean headers at row 0
 *   2. Allstate portal download — metadata rows before header,
 *      header contains "Sub-Producer Name"
 *
 * Exports a single entry-point:
 *   parseNB(fileBuffer, fileName, spine, targetDate) → ParseResult
 */

import * as XLSX from "xlsx";
import type { ParseResult } from "../types";
import { Spine } from "../spine";

// ─── Helpers ────────────────────────────────────────────────────────────────

const EXCEL_EPOCH_MS = Date.UTC(1899, 11, 30);

/**
 * Convert an unknown cell value to "YYYY-MM-DD" string, or null.
 * Handles Excel serial numbers, Date objects, ISO strings, and US-format dates.
 */
function toDateStr(v: unknown): string | null {
  if (v == null) return null;

  if (typeof v === "number" && v > 1 && v < 3000000) {
    const d = new Date(EXCEL_EPOCH_MS + v * 86_400_000);
    return fmtDate(d);
  }

  if (v instanceof Date) {
    if (isNaN(v.getTime())) return null;
    return fmtDate(v);
  }

  if (typeof v === "string") {
    const s = v.trim();
    if (!s) return null;

    const mdy = s.match(/^(\d{1,2})[/\-](\d{1,2})[/\-](\d{4})$/);
    if (mdy) {
      const d = new Date(Date.UTC(+mdy[3], +mdy[1] - 1, +mdy[2]));
      if (!isNaN(d.getTime())) return fmtDate(d);
    }

    const d = new Date(s);
    if (!isNaN(d.getTime())) return fmtDate(d);
  }

  return null;
}

function fmtDate(d: Date): string {
  const y = d.getUTCFullYear();
  const m = String(d.getUTCMonth() + 1).padStart(2, "0");
  const day = String(d.getUTCDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

function str(v: unknown): string {
  if (v == null) return "";
  return String(v).trim();
}

function toFloat(v: unknown): number {
  if (v == null) return 0;
  const n = typeof v === "number" ? v : parseFloat(String(v));
  return isNaN(n) ? 0 : n;
}

function toInt(v: unknown): number {
  if (v == null) return 0;
  const n = typeof v === "number" ? Math.round(v) : parseInt(String(v), 10);
  return isNaN(n) ? 0 : n;
}

// ─── Sheet reading ──────────────────────────────────────────────────────────

interface SheetRow {
  [col: string]: unknown;
}

/**
 * Read an Excel or CSV buffer and auto-detect the header row by scanning
 * the first `probeRows` rows for a cell containing `markerColumn`.
 */
function readWithHeaderDetection(
  fileBuffer: ArrayBuffer,
  fileName: string,
  markerColumn: string,
  probeRows: number,
): { rows: SheetRow[]; columns: string[] } {
  const isCSV = /\.csv$/i.test(fileName);
  let allRows: unknown[][];

  if (isCSV) {
    allRows = parseCSVRaw(fileBuffer);
  } else {
    const wb = XLSX.read(fileBuffer, { type: "array", cellDates: false });
    const ws = wb.Sheets[wb.SheetNames[0]];
    allRows = XLSX.utils.sheet_to_json(ws, { header: 1, raw: true, defval: null }) as unknown[][];
  }

  // Probe for the header row
  let headerIdx = 0;
  const limit = Math.min(probeRows, allRows.length);
  for (let i = 0; i < limit; i++) {
    const row = allRows[i];
    if (!row) continue;
    const vals = row.filter((c) => c != null).map((c) => String(c).trim());
    if (vals.some((v) => v.includes(markerColumn))) {
      headerIdx = i;
      break;
    }
  }

  const headerRow = (allRows[headerIdx] || []).map((c) => str(c));
  const columns = headerRow.map((c) => c.trim());

  const rows: SheetRow[] = [];
  for (let i = headerIdx + 1; i < allRows.length; i++) {
    const raw = allRows[i];
    if (!raw) continue;
    const obj: SheetRow = {};
    let hasData = false;
    for (let j = 0; j < columns.length; j++) {
      const val = raw[j] ?? null;
      obj[columns[j]] = val;
      if (val != null && String(val).trim() !== "") hasData = true;
    }
    if (hasData) rows.push(obj);
  }

  return { rows, columns };
}

function parseCSVRaw(buffer: ArrayBuffer): unknown[][] {
  const text = new TextDecoder("utf-8").decode(buffer);
  const lines = text.split(/\r?\n/);
  const result: unknown[][] = [];

  for (const line of lines) {
    if (!line.trim()) continue;
    const row: unknown[] = [];
    let i = 0;
    while (i < line.length) {
      if (line[i] === '"') {
        let val = "";
        i++;
        while (i < line.length) {
          if (line[i] === '"' && i + 1 < line.length && line[i + 1] === '"') {
            val += '"';
            i += 2;
          } else if (line[i] === '"') {
            i++;
            break;
          } else {
            val += line[i];
            i++;
          }
        }
        row.push(val);
        if (i < line.length && line[i] === ",") i++;
      } else {
        let val = "";
        while (i < line.length && line[i] !== ",") {
          val += line[i];
          i++;
        }
        row.push(val.trim());
        if (i < line.length) i++;
      }
    }
    result.push(row);
  }
  return result;
}

// ─── Agent resolution ───────────────────────────────────────────────────────

/**
 * Resolve agent from an NB row using dual-field fallback:
 *   Primary:  "Sub-Producer Name"
 *   Fallback: "Bind ID Name" (used when Sub-Producer Name is blank, e.g. rewrites)
 * Returns the canonical agent name, or null if unresolvable.
 */
function resolveAgentWithFallback(
  row: SheetRow,
  spine: Spine,
  logs: string[],
): string | null {
  // Try primary field: Sub-Producer Name
  const subName = str(row["Sub-Producer Name"]);
  if (subName) {
    const result = spine.resolveAgent(subName);
    if (result !== null) return result;
  }

  // Fallback: Bind ID Name
  const bindName = str(row["Bind ID Name"]);
  if (bindName) {
    const result = spine.resolveAgent(bindName);
    if (result !== null) {
      const subCode = str(row["Sub Producer"]);
      logs.push(
        `[nb] Fallback: resolved '${bindName}' via Bind ID Name (Sub Producer=${subCode})`,
      );
      return result;
    }
  }

  return null;
}

// ─── Exclude cancelled policies ─────────────────────────────────────────────

function excludeCancelled(
  rows: SheetRow[],
  columns: string[],
  fileName: string,
  logs: string[],
): SheetRow[] {
  if (!columns.includes("Disposition Code")) return rows;

  const kept: SheetRow[] = [];
  let cancelledCount = 0;

  for (const row of rows) {
    const disp = str(row["Disposition Code"]).toLowerCase();
    if (disp === "cancelled") {
      cancelledCount++;
    } else {
      kept.push(row);
    }
  }

  if (cancelledCount > 0) {
    const noun = cancelledCount === 1 ? "policy" : "policies";
    logs.push(`[nb] Excluded ${cancelledCount} cancelled ${noun} from ${fileName}`);
  }

  return kept;
}

// ─── Determine date column ──────────────────────────────────────────────────

function determineDateCol(columns: string[]): string {
  if (columns.includes("Date")) return "Date";
  if (columns.includes("Issued Date")) return "Issued Date";
  if (columns.includes("Date Written")) return "Date Written";
  return "Date"; // fallback
}

// ─── Parse (standard NB counting) ──────────────────────────────────────────

interface ParseBasicResult {
  rows: Record<string, unknown>[];
  logs: string[];
}

function parseBasic(
  fileBuffer: ArrayBuffer,
  fileName: string,
  spine: Spine,
  targetDate: string | null,
): ParseBasicResult {
  const logs: string[] = [];
  const { rows, columns } = readWithHeaderDetection(
    fileBuffer,
    fileName,
    "Sub-Producer Name",
    6,
  );
  logs.push(`[nb] Read ${rows.length} data rows from ${fileName}`);

  // Exclude cancelled policies
  let filtered = excludeCancelled(rows, columns, fileName, logs);

  // Determine and normalize date column
  const dateCol = determineDateCol(columns);
  for (const row of filtered) {
    row["__date"] = toDateStr(row[dateCol]);
  }

  // Filter to target date if specified
  if (targetDate) {
    filtered = filtered.filter((r) => r["__date"] === targetDate);
    logs.push(`[nb] Filtered to targetDate=${targetDate}: ${filtered.length} rows`);
  }

  // Resolve agents with fallback
  const resolved: (SheetRow & { Agent: string })[] = [];
  for (const row of filtered) {
    const agent = resolveAgentWithFallback(row, spine, logs);
    if (agent) {
      (row as Record<string, unknown>)["Agent"] = agent;
      resolved.push(row as SheetRow & { Agent: string });
    }
  }
  logs.push(
    `[nb] Resolved ${resolved.length} rows to agents (${filtered.length - resolved.length} unmapped/dropped)`,
  );

  // Group by [Date, Agent]: NBCount = nunique(Policy No), Items = sum(Item Count),
  //   WrittenPremium = sum(Written Premium)
  const groups = new Map<
    string,
    { policies: Set<string>; items: number; premium: number }
  >();

  for (const row of resolved) {
    const key = `${row["__date"]}||${row["Agent"]}`;
    if (!groups.has(key)) {
      groups.set(key, { policies: new Set(), items: 0, premium: 0 });
    }
    const g = groups.get(key)!;
    const policyNo = str(row["Policy No"]);
    if (policyNo) g.policies.add(policyNo);
    g.items += toInt(row["Item Count"]);
    g.premium += toFloat(row["Written Premium"]);
  }

  const aggRows: Record<string, unknown>[] = [];
  for (const [key, g] of groups) {
    const [date, agent] = key.split("||");
    aggRows.push({
      Date: date,
      Agent: agent,
      NBCount: g.policies.size,
      Items: g.items,
      WrittenPremium: Math.round(g.premium * 100) / 100,
    });
  }

  logs.push(`[nb] Aggregated into ${aggRows.length} (Date, Agent) groups`);
  return { rows: aggRows, logs };
}

// ─── Parse Auto (Standard Auto NB) ─────────────────────────────────────────

interface ParseAutoResult {
  rows: Record<string, unknown>[];
  logs: string[];
}

function parseAuto(
  fileBuffer: ArrayBuffer,
  fileName: string,
  spine: Spine,
  targetDate: string | null,
): ParseAutoResult {
  const logs: string[] = [];
  const { rows, columns } = readWithHeaderDetection(
    fileBuffer,
    fileName,
    "Sub-Producer Name",
    6,
  );

  // 1. Normalize date column
  const dateCol = determineDateCol(columns);
  for (const row of rows) {
    row["__date"] = toDateStr(row[dateCol]);
  }

  let filtered = rows;
  if (targetDate) {
    filtered = filtered.filter((r) => r["__date"] === targetDate);
  }

  // 2. Filter strictly for Standard Auto
  if (columns.includes("Product")) {
    filtered = filtered.filter((r) => str(r["Product"]).toLowerCase() === "standard auto");
    logs.push(`[nb-auto] After Standard Auto filter: ${filtered.length} rows`);
  } else {
    logs.push("[nb-auto] WARNING: 'Product' column not found, skipping Standard Auto filter!");
  }

  // 3. Filter strictly for New Policy Issued
  if (columns.includes("Disposition Code")) {
    filtered = filtered.filter(
      (r) => str(r["Disposition Code"]).toLowerCase() === "new policy issued",
    );
    logs.push(`[nb-auto] After New Policy Issued filter: ${filtered.length} rows`);
  } else {
    logs.push(
      "[nb-auto] WARNING: 'Disposition Code' column not found, skipping New Policy filter!",
    );
  }

  // 4. Resolve agents (fallback to "Other" instead of dropping)
  for (const row of filtered) {
    const agent = resolveAgentWithFallback(row, spine, logs);
    (row as Record<string, unknown>)["Agent"] = agent ?? "Other";
  }

  if (filtered.length === 0) {
    logs.push(`[nb-auto] No Standard Auto New Policies found in ${fileName}`);
    return { rows: [], logs };
  }

  // 5. Deduplicate on Policy No (keep first occurrence)
  if (columns.includes("Policy No")) {
    const initialLen = filtered.length;
    const seen = new Set<string>();
    const deduped: SheetRow[] = [];
    for (const row of filtered) {
      const policyNo = str(row["Policy No"]);
      if (!policyNo || !seen.has(policyNo)) {
        if (policyNo) seen.add(policyNo);
        deduped.push(row);
      }
    }
    const removedCount = initialLen - deduped.length;
    if (removedCount > 0) {
      logs.push(`[nb-auto] Excluded ${removedCount} duplicate policy rows from ${fileName}`);
    }
    filtered = deduped;
  } else {
    logs.push(
      "[nb-auto] WARNING: 'Policy No' column not found, skipping policy number deduplication!",
    );
  }

  // 6. Aggregate by Date and Agent → NBAutoCount (count), NBAutoItems (sum of Item Count)
  const groups = new Map<string, { count: number; items: number }>();
  for (const row of filtered) {
    const key = `${str(row["__date"])}||${str((row as Record<string, unknown>)["Agent"])}`;
    if (!groups.has(key)) groups.set(key, { count: 0, items: 0 });
    const g = groups.get(key)!;
    g.count += 1;
    g.items += toInt(row["Item Count"]);
  }

  const aggRows: Record<string, unknown>[] = [];
  let totalPolicies = 0;
  let totalItems = 0;
  for (const [key, g] of groups) {
    const [date, agent] = key.split("||");
    aggRows.push({
      Date: date,
      Agent: agent,
      NBAutoCount: g.count,
      NBAutoItems: g.items,
    });
    totalPolicies += g.count;
    totalItems += g.items;
  }

  logs.push(
    `[nb-auto] Standard Auto NB parsed: ${totalPolicies} policies, ${totalItems} items from ${fileName}`,
  );
  return { rows: aggRows, logs };
}

// ─── Main entry point ───────────────────────────────────────────────────────

/**
 * Parse an NB file — runs BOTH standard NB counting and Standard Auto
 * aggregation, merging the results into a single ParseResult.
 *
 * Merged rows contain:
 *   { Date, Agent, NBCount, Items, WrittenPremium, NBAutoCount?, NBAutoItems? }
 */
export function parseNB(
  fileBuffer: ArrayBuffer,
  fileName: string,
  spine: Spine,
  targetDate: string | null,
): ParseResult {
  const logs: string[] = [];

  // Run both parse modes
  const basic = parseBasic(fileBuffer, fileName, spine, targetDate);
  logs.push(...basic.logs);

  const auto = parseAuto(fileBuffer, fileName, spine, targetDate);
  logs.push(...auto.logs);

  // Merge by (Date, Agent)
  const merged = new Map<string, Record<string, unknown>>();

  for (const row of basic.rows) {
    const key = `${row["Date"]}||${row["Agent"]}`;
    merged.set(key, { ...row });
  }

  for (const row of auto.rows) {
    const key = `${row["Date"]}||${row["Agent"]}`;
    const existing = merged.get(key);
    if (existing) {
      existing["NBAutoCount"] = row["NBAutoCount"];
      existing["NBAutoItems"] = row["NBAutoItems"];
    } else {
      merged.set(key, { ...row });
    }
  }

  const rows = Array.from(merged.values());
  logs.push(`[nb] Final merged output: ${rows.length} rows`);

  return {
    type: "nb",
    rows,
    logs,
  };
}

File: automation/src/parsers/nb_parser.py
"""
nb_parser.py — New Business policy data parser.

Handles two formats:
  1. Excel from workbook data sheet (nb_export.xlsx) — clean headers at row 0
  2. Allstate portal download (New Business Details_*.xlsx) — 4 metadata rows,
     then header at row 4, data from row 5.

Output: DataFrame with columns:
  Date, Agent, NBCount (policies), Items, WrittenPremium
"""

import pandas as pd
from pathlib import Path
from datetime import date as date_type

from src.spine import Spine


def parse(file_path: str, spine: Spine, target_date=None, sheet_name=0) -> pd.DataFrame:
    """Parse a single NB file. Auto-detects header row."""
    path = Path(file_path)
    df = _read_with_header_detection(file_path, sheet_name)

    # ── Exclude cancelled policies ──
    df = _exclude_cancelled(df, path.name)

    # The Allstate portal NB file has "Issued Date" — that's the authoritative
    # date. It also has "Date Written" which can differ (bind-date vs issue-date).
    # Workbook exports may just have "Date". Prefer in this order:
    if "Date" in df.columns:
        date_col = "Date"
    elif "Issued Date" in df.columns:
        date_col = "Issued Date"
    elif "Date Written" in df.columns:
        date_col = "Date Written"
    else:
        date_col = "Date"

    if date_col in df.columns:
        df[date_col] = pd.to_datetime(df[date_col], errors="coerce")
        if target_date is not None:
            df = df[df[date_col].dt.date == pd.Timestamp(target_date).date()]

    # Rename for consistency
    if date_col != "Date":
        df = df.rename(columns={date_col: "Date"})

    # Map Sub-Producer Name through Spine, falling back to Bind ID Name
    df["Agent"] = df.apply(lambda row: _resolve_agent_with_fallback(row, spine), axis=1)
    df = df.dropna(subset=["Agent"])

    result = (
        df.groupby(["Date", "Agent"])
        .agg(
            NBCount=("Policy No", "nunique"),
            Items=("Item Count", "sum"),
            WrittenPremium=("Written Premium", "sum"),
        )
        .reset_index()
    )
    result["Items"] = result["Items"].astype(int)

    print(f"[nb_parser] Parsed {len(result)} rows from {path.name}")
    return result


def parse_mtd(file_path: str, spine: Spine, month_start, month_end, sheet_name=0) -> pd.DataFrame:
    """Parse NB data for an entire month to compute MTD items."""
    df = _read_with_header_detection(file_path, sheet_name)

    # ── Exclude cancelled policies ──
    df = _exclude_cancelled(df, Path(file_path).name)

    if "Date" in df.columns:
        date_col = "Date"
    elif "Issued Date" in df.columns:
        date_col = "Issued Date"
    elif "Date Written" in df.columns:
        date_col = "Date Written"
    else:
        date_col = "Date"
    df[date_col] = pd.to_datetime(df[date_col], errors="coerce")
    df = df[(df[date_col] >= pd.Timestamp(month_start)) & (df[date_col] <= pd.Timestamp(month_end))]

    df["Agent"] = df.apply(lambda row: _resolve_agent_with_fallback(row, spine), axis=1)
    df = df.dropna(subset=["Agent"])

    result = (
        df.groupby("Agent")
        .agg(ItemsMTD=("Item Count", "sum"), PremiumMTD=("Written Premium", "sum"))
        .reset_index()
    )
    result["ItemsMTD"] = result["ItemsMTD"].astype(int)
    return result


def parse_downloads(
    downloads_folder: str,
    spine: Spine,
    target_date: date_type | None = None,
) -> pd.DataFrame:
    """Scan Downloads folder for New Business Details files.

    When *target_date* is given, iterate through available NB files and pick
    one whose data date range covers that date.  Falls back to the most
    recently modified file when no date match is found or when *target_date*
    is ``None``.
    """
    folder = Path(downloads_folder)
    files = sorted(folder.glob("New Business Details*.xlsx"),
                   key=lambda f: f.stat().st_mtime, reverse=True)

    if not files:
        print("[nb_parser] No New Business Details files found in Downloads")
        return pd.DataFrame()

    # Try to find a file whose data covers target_date
    if target_date is not None:
        td = pd.Timestamp(target_date).date()
        for nf in files:
            try:
                df = _read_with_header_detection(str(nf))
                # Determine date column
                if "Date" in df.columns:
                    dc = "Date"
                elif "Issued Date" in df.columns:
                    dc = "Issued Date"
                elif "Date Written" in df.columns:
                    dc = "Date Written"
                else:
                    continue
                dates = pd.to_datetime(df[dc], errors="coerce").dropna()
                if len(dates) > 0 and dates.min().date() <= td <= dates.max().date():
                    print(f"[nb_parser] Using: {nf.name} (covers {dates.min().date()} to {dates.max().date()})")
                    return parse(str(nf), spine, target_date)
            except Exception:
                continue
        print(f"[nb_parser] No NB file found covering {td}, using newest")

    nb_file = files[0]
    print(f"[nb_parser] Using: {nb_file.name}")
    return parse(str(nb_file), spine, target_date)


def parse_mtd_downloads(
    downloads_folder: str,
    spine: Spine,
    month_start,
    month_end,
) -> pd.DataFrame:
    """Scan Downloads for most recent NB file and compute MTD."""
    folder = Path(downloads_folder)
    files = sorted(folder.glob("New Business Details*.xlsx"),
                   key=lambda f: f.stat().st_mtime, reverse=True)

    if not files:
        print("[nb_parser] No New Business Details files found for MTD")
        return pd.DataFrame()

    return parse_mtd(str(files[0]), spine, month_start, month_end)


def parse_auto(file_path: str, spine: Spine, target_date=None, sheet_name=0) -> pd.DataFrame:
    """
    Parses NB data from file_path for Quotes & NB report page:
      1. Filters to 'Standard Auto' (Product)
      2. Filters to 'New Policy Issued' (Disposition Code)
      3. Resolves agent name (falls back to 'Other' if unmapped)
      4. Excludes duplicate policy numbers
      5. Returns aggregated count (Date, Agent, NBAutoCount, NBAutoItems)
    """
    path = Path(file_path)
    df = _read_with_header_detection(file_path, sheet_name)

    # 1. Normalize Date column
    if "Date" in df.columns:
        date_col = "Date"
    elif "Issued Date" in df.columns:
        date_col = "Issued Date"
    elif "Date Written" in df.columns:
        date_col = "Date Written"
    else:
        date_col = "Date"

    if date_col in df.columns:
        df[date_col] = pd.to_datetime(df[date_col], errors="coerce")
        if target_date is not None:
            df = df[df[date_col].dt.date == pd.Timestamp(target_date).date()]
    if date_col != "Date":
        df = df.rename(columns={date_col: "Date"})

    # 2. Filter strictly for Standard Auto
    if "Product" in df.columns:
        df = df[df["Product"].fillna("").astype(str).str.strip() == "Standard Auto"]
    else:
        print("[nb_parser] WARNING: 'Product' column not found, skipping Standard Auto filter!")

    # 3. Filter strictly for New Policy Issued
    if "Disposition Code" in df.columns:
        df = df[df["Disposition Code"].fillna("").astype(str).str.strip() == "New Policy Issued"]
    else:
        print("[nb_parser] WARNING: 'Disposition Code' column not found, skipping New Policy filter!")

    # 4. Resolve agent names (fallback to "Other" instead of dropping)
    def resolve_agent(row):
        resolved = _resolve_agent_with_fallback(row, spine)
        if resolved:
            return resolved
        return "Other"

    df["Agent"] = df.apply(resolve_agent, axis=1)

    if len(df) == 0:
        print(f"[nb_parser] No Standard Auto New Policies found in {path.name}")
        return pd.DataFrame(columns=["Date", "Agent", "NBAutoCount", "NBAutoItems"])

    # 5. Exclude duplicate policy numbers
    if "Policy No" in df.columns:
        initial_len = len(df)
        df = df.drop_duplicates(subset=["Policy No"], keep="first")
        removed_count = initial_len - len(df)
        if removed_count > 0:
            print(f"[nb_parser] Excluded {removed_count} duplicate policy rows from {path.name}")
    else:
        print("[nb_parser] WARNING: 'Policy No' column not found, skipping policy number deduplication!")

    # 6. Aggregate by Date and Agent
    df["Item Count"] = df["Item Count"].fillna(0).astype(int)
    result = (
        df.groupby(["Date", "Agent"])
        .agg(
            NBAutoCount=("Policy No", "count"),
            NBAutoItems=("Item Count", "sum")
        )
        .reset_index()
    )

    print(f"[nb_parser] Standard Auto NB parsed: {result['NBAutoCount'].sum()} policies, {result['NBAutoItems'].sum()} items from {path.name}")
    return result


def _resolve_agent_with_fallback(row, spine) -> str | None:
    """
    Resolve agent name from a NB row.
    Primary:  Sub-Producer Name (standard field)
    Fallback: Bind ID Name (used when Sub-Producer Name is blank,
              e.g. for rewrites / Sub Producer code 006)
    """
    import pandas as pd

    # Try primary field
    sub_name = row.get("Sub-Producer Name", "")
    if pd.notna(sub_name) and str(sub_name).strip():
        result = spine.resolve_agent(sub_name)
        if result is not None:
            return result

    # Fallback to Bind ID Name
    bind_name = row.get("Bind ID Name", "")
    if pd.notna(bind_name) and str(bind_name).strip():
        result = spine.resolve_agent(bind_name)
        if result is not None:
            sub_code = row.get("Sub Producer", "")
            print(f"[nb_parser] Fallback: resolved '{bind_name}' via Bind ID Name (Sub Producer={sub_code})")
            return result

    return None


def _read_with_header_detection(file_path: str, sheet_name=0) -> pd.DataFrame:
    """
    Read an NB Excel file, auto-detecting the header row.
    Allstate portal downloads have metadata rows; workbook sheets don't.
    """
    # Try reading raw to check if first cell looks like metadata
    probe = pd.read_excel(file_path, engine="openpyxl", sheet_name=sheet_name,
                          header=None, nrows=6)

    # Look for the header row containing "Sub-Producer Name"
    header_row = 0
    for i in range(min(6, len(probe))):
        row_vals = [str(v).strip() for v in probe.iloc[i] if pd.notna(v)]
        if any("Sub-Producer Name" in v for v in row_vals):
            header_row = i
            break

    df = pd.read_excel(file_path, engine="openpyxl", sheet_name=sheet_name,
                       header=header_row)

    # Clean column names (strip whitespace)
    df.columns = [str(c).strip() for c in df.columns]

    return df


def _exclude_cancelled(df: pd.DataFrame, filename: str = "") -> pd.DataFrame:
    """
    Exclude cancelled policies from NB data.
    The Allstate portal NB file uses 'Disposition Code' to indicate status.
    We only want 'New Policy Issued' — anything marked 'Cancelled' is excluded.
    """
    if "Disposition Code" not in df.columns:
        return df

    cancelled = df["Disposition Code"].str.strip().str.lower() == "cancelled"
    n_cancelled = cancelled.sum()
    if n_cancelled > 0:
        print(f"[nb_parser] Excluded {n_cancelled} cancelled polic{'y' if n_cancelled == 1 else 'ies'} from {filename}")
    return df[~cancelled].copy()
